volume/modules/local/quota.go

```diff
@@ -0,0 +1,139 @@
+package local
+
+import (
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// KILOBYTE represents kilo bytes.
+	KILOBYTE = 1024
+
+	// MEGABYTE represents mega bytes.
+	MEGABYTE = 1024 * KILOBYTE
+
+	//GIGABYTE represents giga bytes.
+	GIGABYTE = 1024 * MEGABYTE
+
+	// TERABYTE represents tera bytes.
+	TERABYTE = 1024 * GIGABYTE
+
+	// QuotaMinID represents the minimize quota id.
+	QuotaMinID = uint32(16777216)
+)
+
+var (
+	lock        sync.Mutex
+	quotaLastID uint32
+	// UseQuota represents use quota or not.
+	UseQuota     = true
+	quotaIDs     = make(map[uint32]uint32)
+	mountPoints  = make(map[uint64]string)
+	bytesPattern = regexp.MustCompile(`(?i)^(-?\d+)([KMGT]B?|B)$`)
+	gquota       = NewQuota()
+)
+
+// BaseQuota defines the quota operation interface.
+type BaseQuota interface {
+	QuotaDriverStart(dir string) (string, error)
+	SetSubtree(dir string, qid uint32) (uint32, error)
+	SetDiskQuota(dir string, size string, quotaID int) error
+	CheckMountpoint(devID uint64) (string, bool, string)
+	GetFileAttr(dir string) uint32
+	SetFileAttr(dir string, id uint32) error
+	SetFileAttrNoOutput(dir string, id uint32)
+	GetNextQuatoID() (uint32, error)
+}
+
+// NewQuota returns a quota instance.
+func NewQuota() BaseQuota {
+	return &GrpQuota{}
+}
+
+// GetDevID returns device id.
+func GetDevID(dir string) (uint64, error) {
+	var st syscall.Stat_t
+	if err := syscall.Stat(dir, &st); err != nil {
+		logrus.Warnf("getDirDev: %s, %v", dir, err)
+		return 0, err
+	}
+	return st.Dev, nil
+}
+
+func toByteSize(s string) uint64 {
```

Comment:
Please add a unit test for this function? @rudyfly 😄

Reply:
And this is similar to the function `ToBytes` in vendor package code.cloudfouncry.org/bytefmt/bytes.go.
We have a plan to remove this vendored package, could you encapsulate them into a package in pkg.
See https://github.com/alibaba/pouch/issues/356